File: static/things/js-to-asc/binaryheap_idiomatic.rs
extern crate wee_alloc;

// Use `wee_alloc` as the global allocator.
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
type Scorer<T> = fn(&T) -> f32;
struct BinaryHeap<T> {
    data: Vec<T>,
    scorer: Scorer<T>,
}

struct Item {
    index: usize,
    score: f32,
}

impl<T> BinaryHeap<T> {
    fn new(scorer: Scorer<T>) -> BinaryHeap<T> {
        BinaryHeap {
            data: Vec::with_capacity(10),
            scorer,
        }
    }

    fn push(&mut self, v: T) {
        self.data.push(v);
        self.bubble_up(self.data.len() - 1);
    }

    fn peek(&self) -> Option<&T> {
        self.data.get(0)
    }

    fn pop(&mut self) -> Option<T> {
        if self.data.len() == 0 {
            return None;
        }
        let result = self.data.swap_remove(0);
        if self.data.len() > 0 {
            self.sink_down(0);
        }
        Some(result)
    }

    fn size(&self) -> usize {
        self.data.len()
    }

    fn get_item(&self, n: usize) -> Option<Item> {
        let item = self.data.get(n)?;
        let score = (self.scorer)(item);
        Some(Item { index: n, score })
    }

    fn bubble_up(&mut self, n: usize) {
        let mut element = self.get_item(n).unwrap();
        while element.index > 0 {
            // Compute the parent element's index, and fetch it.
            let parent_n: usize = (((element.index as f32) - 1.0) / 2.0).floor() as usize;
            let parent = self.get_item(parent_n).unwrap();
            // Swap the elements if the parent is greater.
            if element.score < parent.score {
                self.data.swap(element.index, parent.index);
                element.index = parent.index;
            }
            // Found a parent that is less, no need to move it further.
            else {
                break;
            }
        }
    }

    fn sink_down(&mut self, mut n: usize) {
        // Look up the target element and its score.
        let length = self.data.len();
        let element = self.get_item(n).unwrap();

        loop {
            // Compute the indices of the child elements.
            let child2_n = 2 * n + 1;
            let child1_n = child2_n + 1;
            // This is used to store the new position of the element,
            // if any.
            let mut swap: Option<Item> = None;
            // If the first child exists (is inside the array)...
            if child1_n < length {
                // Look it up and compute its score.
                let child1 = self.get_item(child1_n).unwrap();
                // If the score is less than our element's, we need to swap.
                if child1.score < element.score {
                    swap = Some(child1);
                }
            }
            // Do the same checks for the other child.
            if child2_n < length {
                let child2 = self.get_item(child2_n).unwrap();
                if child2.score
                    < swap
                        .as_ref()
                        .map(|item| item.score)
                        .unwrap_or(element.score)
                {
                    swap = Some(child2);
                }
            }

            // If the element needs to be moved, swap it, and continue.
            if let Some(swap) = swap {
                self.data.swap(n, swap.index);
                n = swap.index
            }
            // Otherwise, we are done.
            else {
                break;
            }
        }
    }
}

static mut INSTANCE: Option<BinaryHeap<f32>> = None;

pub fn init() {
    unsafe {
        INSTANCE = Some(BinaryHeap::new(|v: &f32| *v));
    }
}

pub fn push(v: f32) {
    unsafe { INSTANCE.as_mut() }.map(|i| i.push(v));
}

pub fn pop() -> f32 {
    unsafe { INSTANCE.as_mut() }
        .and_then(|i| i.pop())
        .unwrap_or(-1.0)
}

pub fn size() -> usize {
    unsafe { INSTANCE.as_mut() }.map(|i| i.size()).unwrap_or(0)
}

pub fn peek() -> f32 {
    unsafe { INSTANCE.as_mut() }
        .and_then(|i| i.peek())
        .cloned()
        .unwrap_or(-1.0)
}

File: static/things/js-to-asc/binaryheap_idiomatic.cpp
#include "emscripten.h"
#include "emscripten/bind.h"
#include <cstdint>
#include <optional>
#include <utility>
#include <vector>

using namespace emscripten;

class BinaryHeap {
public:
  BinaryHeap() : data() {}

  void push(float v) {
    this->data.push_back(v);
    this->bubble_up(this->data.size() - 1);
  }

  float peek() { return this->data[0]; }

  float pop() {
    float result = this->data[0];
    float end = this->data.back();
    this->data.pop_back();
    if (this->data.size() > 1) {
      this->data[0] = end;
      this->sink_down(0);
    }
    return result;
  }

  uint32_t size() { return this->data.size(); }

private:
  std::vector<float> data;

  void bubble_up(uint32_t n) {
    float element = this->data[n];
    while (n > 0) {
      uint32_t parent_n =
          static_cast<uint32_t>(floor((static_cast<float>(n) + 1) / 2 - 1));
      float parent = this->data[parent_n];
      if (element < parent) {
        std::swap(this->data[parent_n], this->data[n]);
        n = parent_n;
      } else {
        break;
      }
    }
  }

  void sink_down(uint32_t n) {
    auto length = this->data.size();
    float element = this->data[n];
    while (true) {
      uint32_t child2_n = (n + 1) * 2;
      uint32_t child1_n = child2_n - 1;
      std::optional<uint32_t> swap = {};
      float child1, child2;
      if (child1_n < length) {
        child1 = this->data[child1_n];
        if (child1 < element) {
          swap = std::optional<uint32_t>(child1_n);
        }
      }
      if (child2_n < length) {
        child2 = this->data[child2_n];
        if (child2 < (swap.has_value() ? child1 : element)) {
          swap = std::optional<uint32_t>(child2_n);
        }
      }
      if (swap.has_value()) {
        auto swap_n = swap.value();
        std::swap(this->data[n], this->data[swap_n]);
        n = swap_n;
      } else {
        break;
      }
    }
  }
};

thread_local BinaryHeap instance;

void init() { instance = BinaryHeap(); }

void push(float v) { instance.push(v); }

float pop() { return instance.pop(); }

uint32_t size() { return instance.size(); }

float peek() { return instance.peek(); }

EMSCRIPTEN_BINDINGS(my_module) {
  function("init", &init);
  function("push", &push);
  function("pop", &pop);
  function("size", &size);
  function("peek", &peek);
}